File: 15-Flyweight/src/go/forest.go
package main

// Forest is the Client — plants many trees. Each Tree holds only its
// position; the heavy shared data lives once per TreeType in the factory.
type Forest struct {
	trees   []*Tree
	factory *TreeFactory
}

func NewForest() *Forest {
	return &Forest{factory: NewTreeFactory()}
}

func (f *Forest) PlantTree(x, y int, name, color, texture string) {
	kind := f.factory.GetTreeType(name, color, texture)
	f.trees = append(f.trees, &Tree{x: x, y: y, kind: kind})
}

func (f *Forest) Draw() {
	for _, tree := range f.trees {
		tree.Draw()
	}
}

func (f *Forest) TreeCount() int         { return len(f.trees) }
func (f *Forest) DistinctTypeCount() int { return f.factory.DistinctTypeCount() }

File: 15-Flyweight/src/rust/forest.rs
use crate::tree::Tree;
use crate::tree_factory::TreeFactory;

/// Client — plants many trees. Each `Tree` holds only its position;
/// the heavy shared data lives once per `TreeType` in the factory.
pub struct Forest {
    trees: Vec<Tree>,
    factory: TreeFactory,
}

impl Forest {
    pub fn new() -> Self {
        Forest { trees: Vec::new(), factory: TreeFactory::new() }
    }

    pub fn plant_tree(&mut self, x: i32, y: i32, name: &str, color: &str, texture: &str) {
        let kind = self.factory.get_tree_type(name, color, texture);
        self.trees.push(Tree::new(x, y, kind));
    }

    pub fn draw(&self) {
        for tree in &self.trees {
            tree.draw();
        }
    }

    pub fn tree_count(&self) -> usize {
        self.trees.len()
    }

    pub fn distinct_type_count(&self) -> usize {
        self.factory.distinct_type_count()
    }
}
